src/java.base/share/classes/java/lang/reflect/code/op/ExtendedOp.java

```diff
@@ -3152,40 +3152,92 @@ static Block.Builder lowerTypePattern(Block.Builder endNoMatchBlock, Block.Build
                                                   TypePatternOp tpOp, Value target) {
                 TypeElement targetType = tpOp.targetType();
 
-                Block.Builder nextBlock = currentBlock.block();
-
                 // Check if instance of target type
-                Result p;
-                boolean patternWithPrimitive = false;
-                if (targetType instanceof PrimitiveType tt && target.type() instanceof PrimitiveType st) {
-                    patternWithPrimitive = true;
-                    if (SHORT.equals(st) || CHAR.equals(st)) {
-                        st = INT;
+                Op p = null; // op that perform type check
+                Op c; // op that perform conversion
+                TypeElement s = target.type();
+                TypeElement t = targetType;
+                if (t instanceof PrimitiveType pt) {
+                    if (s instanceof ClassType cs) {
+                        // unboxing conversions
+                        ClassType box;
+                        if (cs.unbox().isEmpty()) { // s not a boxed type
+                            // e.g. Number -> int
+                            box = pt.box().orElseThrow();
+                            p = CoreOp.instanceOf(box, target);
+                        } else {
+                            // e.g. Float -> float
+                            box = cs;
+                        }
+                        c = invoke(MethodRef.method(box, t + "Value", t), target);
+                    } else {
+                        // primitive to primitive conversion
+                        PrimitiveType ps = ((PrimitiveType) s);
+                        if (isNarrowingPrimitiveConv(ps, pt) || isWideningPrimitiveConvThatNeedCheck(ps, pt)
+                                || isWideningAndNarrowingPrimitiveConv(ps, pt)) {
+                            // e.g. int -> byte, narrowing
+                            // e,g. int -> float, widening that need check
+                            // e.g. byte -> char, widening and narrowing
+                            MethodRef mref = convMethodRef(s, t);
+                            p = invoke(mref, target);
+                        }
+                        c = CoreOp.conv(targetType, target);
                     }
-                    String s = capitalize(st.toString());
-                    String t = capitalize(tt.toString());
-                    String mn = "is%sTo%sExact".formatted(s, t);
-                    JavaType exactConversionSupport = JavaType.type(ClassDesc.of("java.lang.runtime.ExactConversionsSupport"));
-                    MethodRef mref = MethodRef.method(exactConversionSupport, mn, BOOLEAN, st);
-                    p = currentBlock.op(invoke(mref, target));
                 } else {
-                    p = currentBlock.op(CoreOp.instanceOf(targetType, target));
+                    // e.g. byte -> Byte, Number -> Double, ...
```

Comment:
I think the primitive -> ref case, e.g., byte -> Byte, is not supported correctly. We need to box, which I believe is unconditional if from valid source. You can have another `else if (s instanceof PrimitiveType ps) { ... }`.

The boxing test is likely passing because the interpreter naturally boxes all primitive values and we don't differentiate between a primitive value and its box. Consider that a pragmatic limitation.